### ISSUE TITLE
 could you tell me about the AM-Softmax + Cross-Entropy (CE) loss function  and how to generate these heatmaps. I would greatly appreciate it if you could share those

### ISSUE THREAD

**Issue by zhaochuanxu** [navidnadery/ser_aftl#1]
could you tell me about the AM-Softmax + Cross-Entropy (CE) loss function  and how to generate these heatmaps. I would greatly appreciate it if you could share those

**Comment by navidnadery** [comment]
Hi @zhaochuanxu
Yeah sure, I've just added the loss functions to the repository recently.
To generate these heatmaps, you should get inside pytorch codes (in my own pc is `/home/navid/.conda/envs/torch/lib/python3.9/site-packages/torchaudio/models/wav2vec2/components.py`) and put the code of drawing attention heatmap after the calculation of the weights of SelfAttention (`weights = torch.nn.functional.softmax(weights, dim=-1)`) as follow:

```python
        import matplotlib.pyplot as plt
        fig, ax = plt.subplots()
        avgweight = weights.sum(1) 
        cax = ax.matshow(avgweight[0], cmap='viridis')
        # Add colorbar to interpret values
        fig.colorbar(cax)
        # Display the heatmap
        plt.xlabel('query')
        plt.ylabel('key')
        plt.title('Attention Weights Heatmap')
        plt.savefig(f"/home/navid/Pictures/Ses01F_script02_2_F034_{l}.png")
```

It will be called when wav2vec feature extractor is called:
```python
features, _ = model.extract_features(waveform)
```

**Comment by zhaochuanxu** [comment]
> 嗨，是的，当然，我最近刚刚将损失函数添加到存储库中。要生成这些热图，您应该进入 pytorch 代码（在我自己的电脑中是 ），并在计算 SelfAttention （） 的权重后将绘制注意力热图的代码如下：`/home/navid/.conda/envs/torch/lib/python3.9/site-packages/torchaudio/models/wav2vec2/components.py``weights = torch.nn.functional.softmax(weights, dim=-1)`
> 
> ```python
>         import matplotlib.pyplot as plt
>         fig, ax = plt.subplots()
>         avgweight = weights.sum(1) 
>         cax = ax.matshow(avgweight[0], cmap='viridis')
>         # Add colorbar to interpret values
>         fig.colorbar(cax)
>         # Display the heatmap
>         plt.xlabel('query')
>         plt.ylabel('key')
>         plt.title('Attention Weights Heatmap')
>         plt.savefig(f"/home/navid/Pictures/Ses01F_script02_2_F034_{l}.png")
> ```
> 
> 当调用 wav2vec 特征提取器时，它将被调用：
> 
> ```python
> features, _ = model.extract_features(waveform)
> ```

Thank you very much for your reply, I wish you a happy life and all the best

**Comment by zhaochuanxu** [comment]
> Hi @zhaochuanxu Yeah sure, I've just added the loss functions to the repository recently. To generate these heatmaps, you should get inside pytorch codes (in my own pc is `/home/navid/.conda/envs/torch/lib/python3.9/site-packages/torchaudio/models/wav2vec2/components.py`) and put the code of drawing attention heatmap after the calculation of the weights of SelfAttention (`weights = torch.nn.functional.softmax(weights, dim=-1)`) as follow:
> 
> ```python
>         import matplotlib.pyplot as plt
>         fig, ax = plt.subplots()
>         avgweight = weights.sum(1) 
>         cax = ax.matshow(avgweight[0], cmap='viridis')
>         # Add colorbar to interpret values
>         fig.colorbar(cax)
>         # Display the heatmap
>         plt.xlabel('query')
>         plt.ylabel('key')
>         plt.title('Attention Weights Heatmap')
>         plt.savefig(f"/home/navid/Pictures/Ses01F_script02_2_F034_{l}.png")
> ```
> 
> It will be called when wav2vec feature extractor is called:
> 
> ```python
> features, _ = model.extract_features(waveform)
> ```

I will try to run it. Thank you very much for your reply. I wish you a happy life and all the best
